rust/src/jitter.rs

```diff
@@ -0,0 +1,368 @@
+}
+
+impl<T: AccountProvider> Jitter<T> {
+    pub fn new(
+        drift_client: DriftClient<T>,
+        jitter: Arc<dyn JitterStrategy + Send + Sync>, 
+    ) -> Arc<Self> {
+
+        let (auction_sender, mut auction_receiver): (Sender<Box<dyn Event>>, Receiver<Box<dyn Event>>) = mpsc::channel(100); 
+
+        let jitter = Arc::new(Jitter {
+            drift_client,
+            perp_params: DashMap::new(),
+            spot_params: DashMap::new(),
+            ongoing_auctions: DashMap::new(),
+            user_filter: None,
+            auction_sender,
+            jitter,
+        });
+
+        let jitter_clone = jitter.clone();
+        // Set up a task to monitor for auction events & bounce them to the trampoline
+        tokio::spawn(async move {
+            while let Some(event) = auction_receiver.recv().await {
+                let _ = jitter_clone.on_auction_sync(event);
+            }
+        });
+
+        jitter
+    }
+
+    // Subscribe to auction events and start listening for them
+    pub async fn subscribe(self: Arc<Self>, url: String) -> JitResult<()>{
+
+        let auction_subscriber_config = AuctionSubscriberConfig {
+            commitment: CommitmentConfig::processed(),
+            resub_timeout_ms: None,
+            url,
+        };
+
+        let mut auction_subscriber = AuctionSubscriber::new(auction_subscriber_config);
+        auction_subscriber.subscribe().await?;
+
+        let auction_sender = self.auction_sender.clone();
+        auction_subscriber.event_emitter.subscribe("auction",  move |event| {
+            let _ = auction_sender.try_send(event);
+        });
+
+        futures::future::pending::<()>().await;  // Keep the event loop alive forever
+        Ok(())
+    }
+
+    // Simple trampoline to the async function we need from the auction event 
+    pub fn on_auction_sync(self: &Arc<Self>, event: Box<dyn Event>) -> JitResult<()> {
+        let jitter = self.clone();
+        tokio::spawn(async move {
+            let _ = jitter.on_auction(event).await;
+        });
+
+        Ok(())
+    }
+
+    // Process the auction event & attempt to fill with JIT if possible
+    pub async fn on_auction(&self, event: Box<dyn Event>) -> JitResult<()> {
+        if let Some(auction) = event.as_any().downcast_ref::<ProgramAccountUpdate<User>>() {
+            log::info!("Auction received");
+
+            let user_pubkey = auction.pubkey.clone();
+            let user = auction.data_and_slot.data.clone();
+            let slot = auction.data_and_slot.slot;
+            let user_stats_key = Wallet::derive_stats_account(&user.authority, &drift_program);
+
+            for order in user.orders {  
+                if order.status != OrderStatus::Open {
+                    continue;
+                }
+
+                if !order.has_auction_price(order.slot, order.auction_duration, slot)? {
```

Comment:
technically you can still jit fill after auction has concluded. so it might be worthwhile to use `order.has_auction` here just in case it crosses your market